go/proxyd/daisychain.go

```diff
@@ -0,0 +1,485 @@
+package proxyd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/gorilla/mux"
+	"github.com/gorilla/websocket"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/rs/cors"
+)
+
+var (
+	// Represents the chain id for the Optimism L2 mainnet
+	MainnetChainId = big.NewInt(10)
+	// Represents the chain id for the Optimism L2 public testnet
+	KovanChainId = big.NewInt(69)
+)
+
+// maxEpoch represents the maximum epoch
+// supported by the daisychain. If the RequestOptions
+// is sent with an epoch larger than this, it is an
+// error
+const maxEpoch = uint(6)
+
+// DaisyChainServer is the main struct representing
+// the daisy chain server. It has a reference to
+// each possible backend
+type DaisyChainServer struct {
+	rpcServer          *http.Server
+	maxBodySize        int64
+	authenticatedPaths map[string]string
+	epoch1             *Backend
```

Comment:
Rather than a big list of epochs like this, can we make this an array of `Backend` objects? That way we can use an array in the TOML config to represent each backend in order.

Reply:
We also won't have 6 epochs on Kovan - likely just one.

Reply:
How do we guarantee that the ordering in the config was correct tho?
I generally agree with the idea that we want to decouple things since different networks will have different configs when it comes to the epochs

Reply:
> How do we guarantee that the ordering in the config was correct tho?

What do you mean? The data structure is a TOML array, so the order is guaranteed. As far as how we guarantee the ordering is correct in the config, we'd have to make sure we enter the correct values into the config when we deploy.

Reply:
My concern here is that an infra provider or an enthusiast sets up the config in the incorrect way so just thinking of how to make it as simple as possible to config. I'm looking for some sort of runtime validation